docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,438 @@
+
+#### As an infrastructure provider author, I would like to build a controller to manage provisioning machines without being restricted to a CRUD API.
+
+#### As an infrastructure provider consumer, I would like to have validation for the provider-specific data I need to give the system to have it provision a machine.
+
+### Implementation Details/Notes/Constraints
+
+#### Machine Controller Role
+The Machine Controller lives within Cluster API manager alongside MachineSet and MachineDeployment controllers. This controller should be the only controller having write permissions to Machine objects. The controller responsibility focuses on state transitions and operations including but not limited to:
+- Manage Machines finalizers.
+- Set Machine OwnerReferences (e.g. to Cluster, MachineSet).
+- Setting Machine.Status.Phase based on the transition conditions described above.
+- Pull data from provider custom resources into the right places (e.g. Machine.Spec.Bootstrap.Data from Machine.Spec.Bootstrap.ConfigRef -> Status.BootstrapData).
+- Update Status fields from provider custom resources.
+
+#### Machine Controller dynamic watchers
+The Machine Controller needs to watch for updates to bootstrap and infrastructure provider specific resources so it can copy bootstrap data/status and infrastructure provider status information to the Machine status. To achieve this, we can use controller-runtime’s `source.Informer` type and client-go’s `dynamicinformer.DynamicSharedInformerFactory`.
+
+When the Machine Controller reconciles a Machine and sees a reference to a provider-specific resource, such as a KubeadmBootstrapConfig or an AWSMachineConfig, it can:
+- Get a dynamic informer for the provider-specific resource.
+- Invoke c.Watch() (where c is a `controller.Controller` instance) with a `source.Informer` for the dynamic informer and a `handler.EnqueueRequestForOwner` configured with Machine as the OwnerType.
+
+#### Object References, Templates, MachineSets and MachineDeployments
+With providerSpec and providerStatus no longer inlined in the Machine resource, we need a way to ensure we can continue to provide at least unique status information for a Machine per bootstrap configuration and per infrastructure configuration.
+
+Each Machine needs a reference to a unique Bootstrap provider-specific resource, and a reference to a unique Infrastructure provider-specific resource. This is easy to do when creating a single Machine, but not so obvious when a MachineSet is creating multiple Machine replicas.
+
+For the MachineSet, the references to the Bootstrap and Infrastructure resources are references to templates. When the MachineSet Controller creates a new Machine replica, it retrieves the Bootstrap and Infrastructure resources referenced by the MachineSet’s Machine spec template and creates copies for that Machine to use exclusively.
```

Comment:
I am strongly against having machineset controller to create other resources than machines. Given machineset is a generic controller, it should not be allowed to create cloud provider specific resources (even if it means to just create CRs from templates). Instead, it should be up to infrastructure/bootstrap provider to create those resources since only those the new controllers can interpret the resources.

This decision is out of scope of the new data model. So I take the paragraph to be a recommendation instead of the expected behavior. Yet, if it is supposed to be this way, I vote for making it optional and allowing consumers to decide which controller(s) will be responsible for generating the resource(s) from template(s).

Reply:
That's an interesting point, and it would allow us to actually give full control of provider-specific CRDs (one of the project goals). What do you think to capture this feedback in an issue after the first iteration of this proposal is implemented?

Reply:
Part of the point of working together on this is to establish common expectations for all provider implementations to ensure they work the same way. So, I think if we're going to say that each provider is going to manage its own objects some of the time, we need to say the provider is responsible for doing it all of the time. Otherwise the MachineSet controller needs a way to know if it should do the copying or not.

That said, I'm not sure why there's an issue with having the MachineSet copy all of the data needed as part of creating a Machine.

Reply:
I agree, providers should make copies of the objects or not (they could use a completely different object type if they wanted), not machine-sets.

Keeping machinesets to just machines allows one to opt-into the distributed controller model (machine-controller, bootstrap-controller, infra-controller).  Users could implement a machine-controller that doesn't need to worry about that state machine in fine detail and would otherwise make the machine-api more portable across various implementations.